include/kitepp/utils.hpp

```diff
@@ -2,48 +2,600 @@
+                    if (auto res = client.Put(path.c_str(), headers,
+                            serializedBody, "application/json")) {
+                        code = res->status;
+                        data = res->body;
+                    } else {
+                        throw libException(
+                            FMT("request failed({0})", res.error()));
+                    }
+                }
+                break;
+            case utils::http::METHOD::DEL:
+                if (auto res = client.Delete(path.c_str(), headers)) {
+                    code = res->status;
+                    data = res->body;
+                } else {
+                    throw libException(
+                        FMT("request failed ({0})", res.error()));
+                }
+                break;
+            default: throw libException("unsupported http method");
+        };
+
+        return { code, data, responseType == CONTENT_TYPE::JSON };
+    };
+
+    utils::http::METHOD method;
+    string path;
+    string authToken;
+    Params body;
+    CONTENT_TYPE contentType = CONTENT_TYPE::NON_JSON;
+    CONTENT_TYPE responseType = CONTENT_TYPE::JSON;
+    string serializedBody;
+};
+} // namespace http
+
+namespace ws::ERROR_CODE {
+const unsigned int NORMAL_CLOSURE = 1000;
+const unsigned int NO_REASON = 1006;
+} // namespace ws::ERROR_CODE
+
+template <class Param>
+void addParam(http::Params& bodyParams, Param& param, const string& fieldName) {
+    static_assert(
+        isOptional<std::decay_t<Param>>::value, "Param must be std::optional");
+    if (param.has_value()) {
+        string fieldValue;
+        if constexpr (!std::is_same_v<typename Param::value_type, string>) {
+            fieldValue = std::to_string(param.value());
+        } else {
+            fieldValue = param.value();
+        }
+        if (param.has_value()) { bodyParams.emplace(fieldName, fieldValue); }
+    }
+};
+
+inline std::vector<string> parseInstruments(const std::string& data) {
+    constexpr char seperator = '\n';
+    std::size_t start = 0;
+    std::size_t end = 0;
     std::vector<std::string> tokens;
-    std::size_t start = 0, end = 0;
-    while ((end = text.find(sep, start)) != std::string::npos) {
-        tokens.emplace_back(text.substr(start, end - start));
+    while ((end = data.find(seperator, start)) != std::string::npos) {
+        string token = data.substr(start, end - start - 1);
+        tokens.emplace_back(token);
+        start = end + 1;
+    };
+    // remove headers
+    tokens.erase(tokens.begin());
+    return tokens;
+};
 
+inline std::vector<string> split(const std::string& text, char seperator) {
+    std::vector<std::string> tokens;
+    std::size_t start = 0;
+    std::size_t end = 0;
+    while ((end = text.find(seperator, start)) != std::string::npos) {
+        tokens.emplace_back(text.substr(start, end - start));
         start = end + 1;
     };
 
     tokens.emplace_back(text.substr(start));
-
     return tokens;
 };
```

Comment:
Parsing CSV fields manually like this is bound to fail at some point, especially in financial data. Someday a company with a comma in the name like `Quora, Inc.` might show up in the instrument file. :)

I would suggest to use a standards compliant CSV parsing lib to parse fields.

Reply:
Makes sense. I have fixed this.